File: src/main/java/com/chaottic/nbt/CompoundTag.java
package com.chaottic.nbt;

import com.chaottic.nbt.number.NumberTag;
import it.unimi.dsi.fastutil.bytes.ByteList;
import it.unimi.dsi.fastutil.doubles.DoubleList;
import it.unimi.dsi.fastutil.floats.FloatList;
import it.unimi.dsi.fastutil.ints.IntList;
import it.unimi.dsi.fastutil.longs.LongList;
import it.unimi.dsi.fastutil.shorts.ShortList;

import java.io.DataInput;
import java.io.IOException;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public sealed class CompoundTag implements Tag permits MutableCompoundTag {
    protected final Map<String, Tag> map;

    protected CompoundTag(Map<String, Tag> map) {
        this.map = map;
    }

    public byte getByte(String name) {
        if (map.containsKey(name) && map.get(name) instanceof NumberTag numberTag) {
            return numberTag.getAsByte();
        }
        return 0;
    }

    public short getShort(String name) {
        if (map.containsKey(name) && map.get(name) instanceof NumberTag numberTag) {
            return numberTag.getAsShort();
        }
        return 0;
    }

    public int getInt(String name) {
        if (map.containsKey(name) && map.get(name) instanceof NumberTag numberTag) {
            return numberTag.getAsInt();
        }
        return 0;
    }

    public long getLong(String name) {
        if (map.containsKey(name) && map.get(name) instanceof NumberTag numberTag) {
            return numberTag.getAsLong();
        }
        return 0L;
    }

    public float getFloat(String name) {
        if (map.containsKey(name) && map.get(name) instanceof NumberTag numberTag) {
            return numberTag.getAsFloat();
        }
        return 0.0F;
    }

    public double getDouble(String name) {
        if (map.containsKey(name) && map.get(name) instanceof NumberTag numberTag) {
            return numberTag.getAsDouble();
        }
        return 0.0D;
    }

    public String getString(String name) {
        if (map.containsKey(name) && map.get(name) instanceof StringTag stringTag) {
            return stringTag.s();
        }
        return "";
    }

    public ByteList getByteList(String name) {
        if (map.containsKey(name) && map.get(name) instanceof ListTag listTag) {
            return listTag.asByteList();
        }
        return ByteList.of();
    }

    public ShortList getShortList(String name) {
        if (map.containsKey(name) && map.get(name) instanceof ListTag listTag) {
            return listTag.asShortList();
        }
        return ShortList.of();
    }

    public IntList getIntList(String name) {
        if (map.containsKey(name) && map.get(name) instanceof ListTag listTag) {
            return listTag.asIntList();
        }
        return IntList.of();
    }

    public LongList getLongList(String name) {
        if (map.containsKey(name) && map.get(name) instanceof ListTag listTag) {
            return listTag.asLongList();
        }
        return LongList.of();
    }

    public FloatList getFloatList(String name) {
        if (map.containsKey(name) && map.get(name) instanceof ListTag listTag) {
            return listTag.asFloatList();
        }
        return FloatList.of();
    }

    public DoubleList getDoubleList(String name) {
        if (map.containsKey(name) && map.get(name) instanceof ListTag listTag) {
            return listTag.asDoubleList();
        }
        return DoubleList.of();
    }

    public static CompoundTag createCompoundTag(DataInput dataInput) throws IOException {
        var map = new HashMap<String, Tag>();

        byte opcode;
        while ((opcode = dataInput.readByte()) != 0) {
            map.put(dataInput.readUTF(), TagTypes.fromOpcode(opcode).createTag(dataInput));
        }

        return new CompoundTag(Collections.unmodifiableMap(map));
    }
}

File: src/main/kotlin/com/chaottic/nbt/CompoundTag.kt
package com.chaottic.nbt

import com.chaottic.nbt.number.NumberTag
import java.io.DataInput
import java.util.*

open class CompoundTag protected constructor(protected open val map: Map<String, Tag>) : Tag {

	// TODO:: Shrink
	fun getByte(name: String): Byte {
		if (map.containsKey(name)) {
			val tag = map[name]

			if (tag is NumberTag) {
				return tag.asByte()
			}
		}

		return 0
	}

	fun getShort(name: String): Short {
		if (map.containsKey(name)) {
			val tag = map[name]

			if (tag is NumberTag) {
				return tag.asShort()
			}
		}

		return 0
	}

	fun getInt(name: String): Int {
		if (map.containsKey(name)) {
			val tag = map[name]

			if (tag is NumberTag) {
				return tag.asInt()
			}
		}

		return 0
	}

	fun getLong(name: String): Long {
		if (map.containsKey(name)) {
			val tag = map[name]

			if (tag is NumberTag) {
				return tag.asLong()
			}
		}

		return 0L
	}

	fun getFloat(name: String): Float {
		if (map.containsKey(name)) {
			val tag = map[name]

			if (tag is NumberTag) {
				return tag.asFloat()
			}
		}

		return 0.0F
	}

	fun getDouble(name: String): Double {
		if (map.containsKey(name)) {
			val tag = map[name]

			if (tag is NumberTag) {
				return tag.asDouble()
			}
		}

		return 0.0
	}

	fun getString(name: String): String {
		if (map.containsKey(name)) {
			val tag = map[name]

			if (tag is StringTag) {
				return tag.s
			}
		}

		return ""
	}

	fun getListTag(name: String, opcode: Byte): ListTag {
		if (map.containsKey(name)) {
			val tag = map[name]

			if (tag is ListTag && tag.opcode == opcode) {
				return tag
			}
		}

		return ListTag.empty
	}

	fun getCompoundTag(name: String): CompoundTag {
		if (map.containsKey(name)) {
			val tag = map[name]

			if (tag is CompoundTag) {
				return tag
			}
		}

		return empty
	}

	companion object : TagType<CompoundTag> {
		@JvmStatic
		private val empty = CompoundTag(emptyMap())

		override fun createTag(dataInput: DataInput): CompoundTag {
			val map = HashMap<String, Tag>()

			var opcode: Byte
			while (dataInput.readByte().also { opcode = it } != 0.toByte()) {
				map[dataInput.readUTF()] = TagType.getFromOpcode(opcode).createTag(dataInput)
			}

			return CompoundTag(Collections.unmodifiableMap(map))
		}
	}
}
